### DISCUSSION TITLE
Support umlauts (ö ä ü) in the domain name

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#993] [Anything Else]
Support umlauts (ö ä ü) in the domain name

[request via twitter](https://twitter.com/NBvGersdorff/status/1389648972227268611)

**Update:** Umlauts and other special characters (including numbers, dots and slashes) are now supported so you can add international domain names to Plausible too.

**Comment by masteradhoc** [comment]
+1 request for this issue

**Comment by tobiasgr** [comment]
+1 from here, but with æ,ø,å ;)... what I do in my own system is to convert to punycode in the setter, and convert back from punycode in the getter. And all matches are made with the punycode version of a domain...

**Comment by vinibrsl** [comment]
An easy solution would be to transform all domains to Punycode on creation. This wouldn't require any data migration, as `plausible.io` is `plausible.io` in Punycode as well. But `èé.fr`, for instance, is `xn--8cac.fr` in Punycode. This can be done using a library we already have: `:idna.encode/2`.

This enables support for International Domain Names (IDNs), but two UX issues may arise.

1. IDNs would be displayed as Punycode (`xn--`) unless all client-side + e-mail references are changed

```diff
     <h2 class="text-xl font-bold dark:text-gray-100">Waiting for first pageview</h2>
-    <h2 class="text-xl font-bold dark:text-gray-100">on <%= @site.domain %></h2>
+    <h2 class="text-xl font-bold dark:text-gray-100">on <%= Plausible.Site.IDN.decode(@site.domain) %></h2>
```

2. URLs would use Punycode, e.g. `https://plausible.io/xn--80andih6b.com`

What are your thoughts on this, @ukutaht?

**Reply by metmarkosaric** [reply]
nice one @vinibrsl! i don't see any issues with nr 2. it might be a bigger issue if all these domains are displayed in Punycode in the UI itself so we should try and avoid that. does changing all these client-side + e-mail references add a lot of extra work?

**Reply by vinibrsl** [reply]
It shouldn't be complex, but it requires a bit of effort, as there must be lots of references to domain names in the UI (correct me if I'm wrong). In any case I created a draft pull request so we can start hacking on this https://github.com/plausible/analytics/pull/2034

**Reply by metmarkosaric** [reply]
ok great! Uku can tell you more about the domain name references as i don't know about the coding side of things. but this will certainly be a great improvement to Plausible and make us more user friendly to people around the world who need non-english characters so it's worth doing right. thanks!

**Reply by ukutaht** [reply]
> An easy solution would be to transform all domains to Punycode on creation

Hmm I don't understand what problem this solves. I would think that the browser, HTTP, Elixir, Postgres and Clickhouse all support UTF-8 as a character encoding, which already supports special characters. Before we can evaluate Punycode as a solution, we must first consider what problem we're actually solving. Are any of the underlying technologies limited to ASCII?

**Reply by ukutaht** [reply]
If not, then what is limiting us from creating domains with special characters at the moment?

**Reply by masteradhoc** [reply]
@ukutaht Currently  your not allowing us to add this.
Example: äö-test.li
![image](https://user-images.githubusercontent.com/6242098/180426870-ec1d975d-e411-4bac-9488-3b36d97a0eff.png)

As @vinibrsl wrote we have to currently manually translate it punycode compatible to: xn---test-fra0k.li
then it works :)

But its not readable at all and misleading to our clients.

**Reply by ukutaht** [reply]
Yes I agree that it's currently not allowed in the UI, that's a problem, and we want to fix it.

I wanted to continue the conversation about the technical implementation that @vinibrsl proposed. The recommendation was to encode to Punycode when storing the domain and to decode it when pulling it out of the database. My question is: why is Punycode needed here? Can't we just store the domain as utf-8 which already supports all special characters you might want to use in a domain name?

**Reply by kthy** [reply]
Storing as Punycode makes it easier for you to tell the difference between аlpha.com (`xn--lpha-43d.com`) and alpha.com (`alpha.com`) internally.

**Comment by vinibrsl** [comment]
Hey everyone! #2034 adding international domain names support has been merged 🙌 

It adds support for any Unicode character in the "letter" category. In other words, a letter in any language. Plus numbers, dots, and slashes.

**Reply by metmarkosaric** [reply]
nice one @vinibrsl! so it's live already now? i still get the same error

**Reply by vinibrsl** [reply]
I'll deploy it today. Should be live in an hour :) @metmarkosaric

**Reply by metmarkosaric** [reply]
cool! i'll prepare the announcement

**Comment by tobiasgr** [comment]
💪
